servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherCache.java

```diff
@@ -0,0 +1,257 @@
+                                lockRemoveFromMap();
+                                subscriber.onError(t);
+                            }
+
+                            @Override
+                            public void onComplete() {
+                                lockRemoveFromMap();
+                                subscriber.onComplete();
+                            }
+
+                            private void lockRemoveFromMap() {
+                                synchronized (publisherCache) {
+                                    publisherCache.remove(key, item2);
+                                }
+                            }
+                        });
+
+                item2.publisher = multicastStrategy.apply(newPublisher)
+                        .liftSync(subscriber -> new Subscriber<T>() {
+                            @Override
+                            public void onSubscribe(Subscription subscription) {
+                                subscriber.onSubscribe(new Subscription() {
+                                    @Override
+                                    public void request(long n) {
+                                        subscription.request(n);
+                                    }
+
+                                    @Override
+                                    public void cancel() {
+                                        synchronized (publisherCache) {
+                                            subscription.cancel();
+                                        }
+                                    }
+                                });
+                            }
+
+                            @Override
+                            public void onNext(@Nullable T next) {
+                                subscriber.onNext(next);
+                            }
+
+                            @Override
+                            public void onError(Throwable t) {
+                                try {
+                                    subscriber.onError(t);
+                                } finally {
+                                    lockRemoveFromMap();
+                                }
+                            }
+
+                            @Override
+                            public void onComplete() {
+                                try {
+                                    subscriber.onComplete();
+                                } finally {
+                                    lockRemoveFromMap();
+                                }
+                            }
+
+                            private void lockRemoveFromMap() {
+                                synchronized (publisherCache) {
+                                    publisherCache.remove(key, item2);
+                                }
+                            }
+                        });
+                return item2.publisher;
+            }
+        });
+    }
+
+    private static final class Holder<T> {
```

Comment:
I _think_ this is here so that we can add it to the hashmap and avoid some jumping through hoops to manage the reference equality stuff for `HashMap`. A quick code comment might be helpful for future readers.